stretch_core/stretch_core/stretch_driver.py

```diff
@@ -108,6 +112,62 @@ def set_mobile_base_velocity_callback(self, twist):
         self.angular_velocity_radps = twist.angular.z
         self.last_twist_time = self.get_clock().now()
         self.robot_mode_rwlock.release_read()
+    
+    def set_robot_streaming_position_callback(self, msg):
+        self.robot_mode_rwlock.acquire_read()
+        if not self.streaming_position_activated:
+            self.get_logger().error('Streaming position is not activated.'
+                                    ' Please activate streaming position to receive command to joint_position_cmd.')
+            self.robot_mode_rwlock.release_read()
+            return
+        
+        if not self.robot_mode in ['position', 'navigation']:
+            self.get_logger().error('{0} must be in position or navigation mode with streaming_position activated ' 
+                                    'enabled to receive command to joint_position_cmd. '
+                                    'Current mode = {1}.'.format(self.node_name, self.robot_mode))
+            self.robot_mode_rwlock.release_read()
+            return
+
+        
+        if self.get_clock().now() - self.streaming_controller_lt.last_update_time> 5:
+            if STREAMING_POSITION_DEBUG:
+                print('Reset Streaming position looptimer after 5s no message received.')
+            self.streaming_controller_lt.reset()
+        qpos = msg.data
+        self.move_to_position(qpos)
+        self.robot_mode_rwlock.release_read()
+        self.streaming_controller_lt.update()
+    
+    def move_to_position(self, qpos):
+        try:
+            try:
+             Idx = get_Idx(self.robot.params['tool'])
+            except UnsupportedToolError:
+                self.get_logger().error('Unsupported tool for streaming position control.')
+            if len(qpos) != Idx.num_joints:
+                self.get_logger().error('Received qpos does not match the number of joints in the robot')
+                return
+            self.robot.arm.move_to(qpos[Idx.ARM])
+            self.robot.lift.move_to(qpos[Idx.LIFT])
+            self.robot.end_of_arm.move_to('wrist_yaw', qpos[Idx.WRIST_YAW])
+            if 'wrist_pitch' in self.robot.end_of_arm.joints:
+                self.robot.end_of_arm.move_to('wrist_pitch', qpos[Idx.WRIST_PITCH])
+            if 'wrist_roll' in self.robot.end_of_arm.joints:
+                self.robot.end_of_arm.move_to('wrist_roll', qpos[Idx.WRIST_ROLL])
+            self.robot.head.move_to('head_pan', qpos[Idx.HEAD_PAN])
+            self.robot.head.move_to('head_tilt', qpos[Idx.HEAD_TILT])
+            if abs(qpos[Idx.BASE_TRANSLATE]) > 0.0 and abs(qpos[Idx.BASE_ROTATE]) > 0.0 and self.robot_mode != 'position':
+                self.get_logger().error('Cannot move base in both translation and rotation at the same time in position mode')
+            elif abs(qpos[Idx.BASE_TRANSLATE]) > 0.0 and self.robot_mode == 'position':
+                self.robot.base.translate_by(qpos[Idx.BASE_TRANSLATE])
```

Comment:
As mentioned on Slack I dont love using Idx in this way

are dictionaries with strings (ie JointState) still fast enough?